actors/builtin/miner/miner_actor.go

```diff
@@ -2067,6 +2068,295 @@ func (a Actor) RepayDebt(rt Runtime, _ *abi.EmptyValue) *abi.EmptyValue {
+		sectorsDataSpec = append(sectorsDataSpec, &market.SectorDataSpec{
+			SectorType: sectorInfo.SealProof,
+			DealIDs:    update.Deals,
+		})
+	}
+
+	builtin.RequireParam(rt, len(validatedUpdates) > 0, "no valid updates")
+
+	// Errors past this point cause the ProveReplicaUpdates call to fail (no more skipping sectors)
+
+	dealWeights := requestDealWeights(rt, sectorsDeals)
+	builtin.RequirePredicate(rt, len(dealWeights.Sectors) == len(validatedUpdates), exitcode.ErrIllegalState,
+		"deal weight request returned %d records, expected %d", len(dealWeights.Sectors), len(validatedUpdates))
+
+	unsealedSectorCIDs := requestUnsealedSectorCIDs(rt, sectorsDataSpec...)
+	builtin.RequirePredicate(rt, len(unsealedSectorCIDs) == len(validatedUpdates), exitcode.ErrIllegalState,
+		"unsealed sector cid request returned %d records, expected %d", len(unsealedSectorCIDs), len(validatedUpdates))
+
+	type updateWithDetails struct {
+		update            *ReplicaUpdate
+		sectorInfo        *SectorOnChainInfo
+		dealWeight        market.SectorWeights
+		unsealedSectorCID cid.Cid
+	}
+
+	// Group declarations by deadline
+	declsByDeadline := map[uint64][]*updateWithDetails{}
+	var deadlinesToLoad []uint64
+	for i, updateWithSectorInfo := range validatedUpdates {
+		if _, ok := declsByDeadline[updateWithSectorInfo.update.Deadline]; !ok {
+			deadlinesToLoad = append(deadlinesToLoad, updateWithSectorInfo.update.Deadline)
+		}
+		declsByDeadline[updateWithSectorInfo.update.Deadline] = append(declsByDeadline[updateWithSectorInfo.update.Deadline], &updateWithDetails{
+			update:            updateWithSectorInfo.update,
+			sectorInfo:        updateWithSectorInfo.sectorInfo,
+			dealWeight:        dealWeights.Sectors[i],
+			unsealedSectorCID: unsealedSectorCIDs[i],
+		})
+	}
+
+	rewRet := requestCurrentEpochBlockReward(rt)
+	powRet := requestCurrentTotalPower(rt)
+
+	succeededSectors := bitfield.New()
+	var st State
+	rt.StateTransaction(&st, func() {
+		deadlines, err := st.LoadDeadlines(store)
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load deadlines")
+
+		newSectors := make([]*SectorOnChainInfo, len(validatedUpdates))
+		for _, dlIdx := range deadlinesToLoad {
+			deadline, err := deadlines.LoadDeadline(store, dlIdx)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load deadline %d", dlIdx)
+
+			partitions, err := deadline.PartitionsArray(store)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load partitions for deadline %d", dlIdx)
+
+			quant := st.QuantSpecForDeadline(dlIdx)
+
+			for _, updateWithDetails := range declsByDeadline[dlIdx] {
+				updateProofType, err := updateWithDetails.sectorInfo.SealProof.RegisteredUpdateProof()
+				builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "couldn't load update proof type")
+
+				err = rt.VerifyReplicaUpdate(
+					proof.ReplicaUpdateInfo{
+						UpdateProof:          updateProofType,
```

Comment:
So most other places take this as a param to the method (see WindowPoSt), even though there's only one correct value, and it can be determined at runtime. 

Is there a reason for that pattern? Should the proof type be an input to this method (that we then verify)?

Reply:
Yes please include it.  This is so that proofs can do versioning easily.

Reply:
The ffi itself usually takes these from the runtime methods.

Reply:
Will do -- should the provided proof type be validated that it's `sectorInfo.SealProof.RegisteredUpdateProof()`?

Reply:
After an offline discussion (that also cleared up some miscommunication), the decision is to include the proof type in the params to `ProveReplicaUpdates`. At the cost of one byte, our future selves get the ability to much more easily support multiple valid UpdateProof types at the same time.